stdlib/LinearAlgebra/src/eigen.jl

```diff
@@ -647,9 +647,34 @@ function Base.isequal(A::Eigen, B::Eigen)
 end
 
 # Conversion methods
-
 ## Can we determine the source/result is Real?  This is not stored in the type Eigen
-AbstractMatrix(F::Eigen) = F.vectors * Diagonal(F.values) / F.vectors
+AbstractMatrix(F::Eigen) = F.are_vectors_unitary ?
+                             (F.vectors * Diagonal(F.values) * (F.vectors')) :
+                             (F.vectors * Diagonal(F.values) / F.vectors)
 AbstractArray(F::Eigen) = AbstractMatrix(F)
 Matrix(F::Eigen) = Array(AbstractArray(F))
 Array(F::Eigen) = Matrix(F)
+
+# Operations on Eigen
+# Only operations that can be efficiently implemented or conserve the eigen type are implemented
+adjoint(E::Eigen) = Eigen(adjoint.(E.values),
+                          copy(adjoint(E.are_vectors_unitary ? (E.vectors') : inv(E.vectors))),
+                          E.are_vectors_unitary)
+transpose(E::Eigen) = Eigen(transpose.(E.values),
+                            copy(transpose(E.are_vectors_unitary ? (E.vectors') : inv(E.vectors))),
+                            E.are_vectors_unitary && (eltype(E) <: Real))
+
+det(A::Eigen) = prod(A.values)
+tr(E::Eigen) = sum(E.values)
+
+*(E::Eigen, x::AbstractVector) = E.vectors * (E.values .* (E.are_vectors_unitary ? (E.vectors'x) : (E.vectors \ x)))
+*(x::Adjoint{<:Any, <:AbstractVector}, E::Eigen) = E.are_vectors_unitary ?
+                                                   ((x*(E.vectors)) * Diagonal(E.values)) * (E.vectors') :
+                                                   ((x*(E.vectors)) * Diagonal(E.values)) / E.vectors
+mul!(x, E::Eigen, y::AbstractVector) = copyto!(x, E*y)
+
+inv(E::Eigen) = Eigen(map(inv, E.values), copy(E.vectors), E.are_vectors_unitary)
+for f in _matrix_functions
+    @eval Base.$f(E::Eigen) = Eigen(map($f, E.values), copy(E.vectors), E.are_vectors_unitary)
+end
+^(E::Eigen, x) = Eigen(map(λ -> λ^x, E.values), copy(E.vectors), E.are_vectors_unitary)
```

Comment:
Why do we copy the vectors matrix. Wouldn't a shallow copy of `Eigen` be sufficient?

Reply:
I thought there would be issues to sharing the eigenvalues if somebody wants to modify them, but OK let's just consider them as read only and I can avoid the copy

Reply:
Hard to tell what is right (I tend slightly towards avoiding the copy), so that means we need to document the behaviour.